### ISSUE TITLE
JAX IMPALA agent does not allow hk.dynamic_unroll.

### ISSUE THREAD

**Issue by ethanluoyc** [deepmind/acme#111]
When changing the unroll function of the example agent to use dynamic_unroll, this will result in the following error:

> TypeError: scan carry output and input must have same type structure, got PyTreeDef(tuple, [PyTreeDef(namedtuple[<class 'haiku._src.recurrent.LSTMState'>], [*,*]),PyTreeDef(namedtuple[<class 'haiku._src.stateful.InternalState'>], [PyTreeDef(None, []),PyTreeDef(<class 'collections.defaultdict'>[(<class 'dict'>, ())], []),PyTreeDef(None, [])])]) and PyTreeDef(tuple, [PyTreeDef(namedtuple[<class 'tensorflow.python.saved_model.nested_structure_coder.LSTMState'>], [*,*]),PyTreeDef(namedtuple[<class 'haiku._src.stateful.InternalState'>], [PyTreeDef(None, []),PyTreeDef(<class 'collections.defaultdict'>[(<class 'dict'>, ())], []),PyTreeDef(None, [])])]).

I think the reason that this occurs is that when using the dataset iterator, the hk.LSTMState in the extra_spec is not preserved. This causes jax.scan to error since it requires the type of the initial_state to match the type of the carry. In this case, the initial_state has type tensorflow.python.saved_model.nested_structure_coder.LSTMState while the carry will have type hk.LSTMState

_Notes_
- the signature returned from the replay client is not respecting the type

```python
client = reverb_client.Client(address)
info = client.server_info(-1)
print(info[replay_table_name].signature) # will give tf.python.saved_model.LSTMState
```

- In https://github.com/deepmind/reverb/blob/master/reverb/server.py#L148, the signature gets encoded  into signature_proto_str which is what gives the type above. The client gets information about the table which returns the encoded signature, but this is not mapped back to the original type specified in the extra_spec. It seems that when deserializing the proro string in https://github.com/deepmind/reverb/blob/master/reverb/client.py#L463 does not actually restore the original type.
- A simple repro for this. Looks like an incompatibility of hk.LSTM with the nested_structure_coder
```python

from tensorflow.python.saved_model import nested_structure_coder
import tensorflow as tf
import haiku as hk

tf.__version__ # 2.4.1

spec = {
    'core': hk.LSTMState(tf.TensorSpec(shape=tf.TensorShape((1,2))), tf.TensorSpec(shape=tf.TensorShape((1,2))))}

# Encode into proto
proto = nested_structure_coder.StructureCoder().encode_structure(spec)
# Decode 
type(nested_structure_coder.StructureCoder().decode_proto(proto)['core'])
# Gives tensorflow.python.saved_model.nested_structure_coder.LSTMState
```
The error seems to be that hk.LSTM is a NamedTuple and the decoder is not reconstructing that. Usual tuples work fine however. Apparently, this is documented 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/saved_model/nested_structure_coder.py#L226

 I guess solutions to this problem would be
1. JAX not erroring when the initial carry has a type mismatch during scan. One possibility is for JAX to be more graceful in checking the types for NamedTuple by allowing equality between two different NamedTuples with the same `typename` and `fields`.
2. Perform additional step to map back to the original signature, this can be a little bit tricky since the client does not require a signature at construction so it may be difficult to propagate that information forward. I am not really in favour of this idea since there may be performance implications.
3. Handle this special case in haiku.dynamic_unroll. I am not sure if this is in general possible but just my third opinion.

Depending on what the solution is, maybe I can cross-post this to the respective repo. Do the Acme maintainers have some ideas about this in mind?

Thanks!

**Comment by nikolamomchev** [comment]
@ethanluoyc sorry for the silence on this. I wonder if you managed to find a solution?

TBH, I didn't fully understand the problem, but it sounds like it's not specific to Acme, rather to how you are building the networks passed to Acme? I.e. it seems to me that it's rather related to Haiku and not Acme.

**Comment by ethanluoyc** [comment]
I think in the end I just did a manual cast of my state back to LSTMState. It's a hacky workaround but it works :D

https://github.com/ethanluoyc/magi/blob/2ef2ba60989a55ccf8c90ba74c8e712fe301e2fa/magi/agents/impala/agent_test.py#L76

IIRC, I think the problem is that when reverb unpacks structs from the server, it will only preserve structural identity. Imagine a named tuple A(a:int, b:float). Then reverb samples will return B(a:int, b:float) where A and B are structurally the same but they are not the same type, i.e. type(A) != type(B). This would cause hk.static_unroll to fail some internal checks, but you are right it's not a problem specific to Acme but rather an issue that I found when modifying the IMPALA agent to use dynamic_unroll instead of static_unroll. static_unroll takes too long to compile in my case as I am compiling sequences of >100 in steps..
